File: 0001-0500/0238.Product of Array Except Self.rb
# @param {Integer[]} nums
# @return {Integer[]}
def product_except_self(nums)
    res = [1] * nums.size
    (1...nums.size).each { |i| res[i] = res[i - 1] * nums[i - 1] }
    right = nums[-1]
    (nums.size - 2).downto(0).each do |i|
        res[i] *= right
        right *= nums[i]
    end
    res
end

File: 0001-0500/0238.Product of Array Except Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = [1] * len(nums)
        for i in xrange(1, len(nums)):
            res[i] = res[i - 1] * nums[i - 1]
        right = 1
        for i in xrange(len(nums) - 2, -1, -1):
            right *= nums[i + 1]
            res[i] *= right
        return res
